Add documentation on debugging OpenCL applications with Pocl

franz opened the pull request:
GDB part needs improvement

pjaaskel commented on doc/sphinx/source/debug.rst:
```diff
@@ -0,0 +1,236 @@
+Debugging OpenCL applications with Pocl
```
Let's shorten it PoCL from now on.

pjaaskel commented on doc/sphinx/source/debug.rst:
```diff
@@ -0,0 +1,236 @@
+========================================
+
+
+There are several ways to debug applications with Pocl,
```
PoCL

pjaaskel commented on doc/sphinx/source/debug.rst:
```diff
@@ -0,0 +1,236 @@
+Downsides:
+  * requires rebuilding both the application and pocl;
+  * the application and pocl's runtime code are compiled with sanitizer,
+    but ATM the kernels are not compiled with sanitizer
```
at the moment

pjaaskel commented on doc/sphinx/source/debug.rst:
```diff
@@ -0,0 +1,236 @@
+    but ATM the kernels are not compiled with sanitizer
+
+Upsides:
+  * much faster than Valgrind
```
Move this sanitizer after the valgrind method if referring to it?

pjaaskel commented on doc/sphinx/source/debug.rst:
```diff
@@ -0,0 +1,236 @@
+-----------------------------------------------
+
+Downsides:
+  * extremely slow esp with computationally intensive kernels
```
especially

pjaaskel commented on doc/sphinx/source/debug.rst:
```diff
@@ -0,0 +1,236 @@
+
+Downsides:
+  * extremely slow esp with computationally intensive kernels
+  * ocassionally false positive reports
```
occassionally

pjaaskel commented on doc/sphinx/source/debug.rst:
```diff
@@ -0,0 +1,236 @@
+-----------------------------------------------
+
+The setup is similar to Valgrind; build pocl with ``-DCMAKE_BUILD_TYPE=Debug``
+and export ``POCL_EXTRA_BUILD_FLAGS="-g -cl-opt-disable"`` before running gdb.
```
...or just add these flags to clbprogram options.

franz commented:
OK, fixed pocl to PoCL etc, plus updated the GDB section with examples.

Stepping inside kernel in GDB is possible, though a little tricky; i've added an example for that too.

pjaaskel commented on doc/sphinx/source/debug.rst:
```diff
@@ -0,0 +1,390 @@
+	307	          k->workgroup ((uint8_t*)arguments, (uint8_t*)&pc,
+
+
+Now it's about to launch the ``dot_product`` kernel for a single workgroup, so let's find the kernel source file::
```
Phew. Quite a bit of hoops to go through. Why not just make the POCL_DEBUG dump the temp file name more explicitly so it can be used in gdb?

franz replied:
We can do that, but it won't change the fact that you have to stop at `k->workgroup`  (or somewhere post-compilation-before-execution). So it would really only save one step, the `info functions kernel_name`. If you think it's worth it, i can add the code.

pjaaskel replied:
Can we use the kernel function name to set the breakpoint (including the PoCL WG naming)?

franz replied:
Yes, that works, and it makes the process significantly simpler. I've updated the docs

pjaaskel commented on lib/CL/clReleaseContext.c:
```diff
@@ -67,7 +67,6 @@ POname(clReleaseContext)(cl_context context) CL_API_SUFFIX__VERSION_1_0
 
       /* see below on why we don't call uninit_devices here anymore */
       --cl_context_count;
-      POCL_ATOMIC_DEC (cl_context_count);
```
Why did you remove the atomic one?

pjaaskel commented:
Fails to build for LLVM 8.
```
FlattenGlobals.cc:117:58: error: ‘class llvm::Function’ has no member named ‘hasOptNone’; did you mean ‘hasSection’?
         if (pocl::Workgroup::isKernelToProcess(*f) && f->hasOptNone())
                                                          ^~~~~~~~~~
                                                          hasSection
```
